internal/provider/resource_gitlab_branch_test.go

```diff
@@ -0,0 +1,150 @@
+                    testAccCheckGitlabBranchAttributes(&b, &testAccGitlabBranchExpectedAttributes{
+                        Name: fmt.Sprintf("example-%[1]d", rInt),
+                    }),
+                ),
+            },
+            // Test ImportState
+            {
+                ResourceName:      "gitlab_branch.this",
+                ImportState:       true,
+                ImportStateVerify: true,
+            },
+            // Update the branch to toggle all the values to their inverse
+            {
+                Config: testAccGitlabBranchConfig(rModifyInt, rSupportInt),
+                Check: resource.ComposeTestCheckFunc(
+                    testAccCheckGitlabBranchExists("gitlab_branch.this", &b),
+                    testAccCheckGitlabBranchAttributes(&b, &testAccGitlabBranchExpectedAttributes{
+                        Name: fmt.Sprintf("example-%[1]d", rModifyInt),
+                    }),
+                ),
+            },
+        },
+    })
+}
+
+func testAccCheckGitlabBranchExists(n string, b *gitlab.Branch) resource.TestCheckFunc {
+    return func(s *terraform.State) error {
+        rs, ok := s.RootModule().Resources[n]
+        if !ok {
+            return fmt.Errorf("Not Found: %s", n)
+        }
+
+        project, branch_name, err := parseTwoPartID(rs.Primary.ID)
+        if err != nil {
+            return fmt.Errorf("Error in Splitting Project and Branch Ids")
+        }
+
+        branch, _, err := testGitlabClient.Branches.GetBranch(project, branch_name)
+        if err != nil {
+            return err
+        }
+
+        if branch.Name == branch_name {
+            *b = *branch
+            return nil
+        }
+        return fmt.Errorf("Branch does not exist")
+    }
+}
+
+type testAccGitlabBranchExpectedAttributes struct {
+    Name string
+}
+
+func testAccCheckGitlabBranchAttributes(b *gitlab.Branch, want *testAccGitlabBranchExpectedAttributes) resource.TestCheckFunc {
+    return func(s *terraform.State) error {
+        if b.Name != want.Name {
+            return fmt.Errorf("got name %q; want %q", b.Name, want.Name)
+        }
+        return nil
+    }
+}
+
+func testAccCheckGitlabBranchDestroy(s *terraform.State) error {
+    var project string
+    var branch_name string
+    for _, rs := range s.RootModule().Resources {
+        if rs.Type == "gitlab_project" {
+            project = rs.Primary.ID
+        } else if rs.Type == "gitlab_branch" {
+            branch_name = rs.Primary.ID
+        }
+    }
+
+    branch, _, err := testGitlabClient.Branches.GetBranch(project, branch_name)
+    if err == nil {
+        if branch != nil {
+            return fmt.Errorf("project branch protection %s still exists", branch)
+        }
+    }
+    if !is404(err) {
+        return err
+    }
+    return nil
+}
+
+func testAccGitlabBranchConfig(rInt int, rSupportInt int) string {
+    return fmt.Sprintf(`
+resource "gitlab_group" "this" {
```

Comment:
Please use `testAccCreateGroups` outside of the TestStep to limit the config here to what's actually being tested :)

Reply:
same goes for the other resources which are not under test.

Reply:
Hm, interesting. Ok